### PR TITLE
Fix const_fn ICE with non-const function pointer

### LINKED ISSUE #56164: ICE when compiling `const fn` that calls a non-const function through a pointer
My minimal reproduction:

```rust
#![feature(const_fn)]

const fn bad(input: fn()) {
    input()
}

fn main() {
    bad(|| {});
}
```

Sidenote: I'm pretty sure this is the smallest working ICE that I've seen in Rust so far.

### PR DESCRIPTION
**Pull request by phansch**
Fixes #56164

### CODE REVIEW

File: src/librustc_mir/transform/qualify_consts.rs [hunk 1]
```diff
@@ -846,143 +846,160 @@ impl<'a, 'tcx> Visitor<'tcx> for Qualifier<'a, 'tcx, 'tcx> {
...
+            match fn_ty.sty {
+                ty::FnDef(def_id, _) => {
+                    callee_def_id = Some(def_id);
+                    match self.tcx.fn_sig(def_id).abi() {
+                        Abi::RustIntrinsic |
+                            Abi::PlatformIntrinsic => {
+                                assert!(!self.tcx.is_const_fn(def_id));
+                                match &self.tcx.item_name(def_id).as_str()[..] {
+                                    | "size_of"
+                                        | "min_align_of"
```
Comment (oli-obk) [inline src/librustc_mir/transform/qualify_consts.rs]: what happened to the indent here?

File: src/librustc_mir/transform/qualify_consts.rs [hunk 2]
```diff
@@ -846,143 +846,160 @@ impl<'a, 'tcx> Visitor<'tcx> for Qualifier<'a, 'tcx, 'tcx> {
...
+                ty::FnPtr(_) => {
+                    if self.mode != Mode::Fn {
+                        let mut err = self.tcx.sess.struct_span_err(
+                            self.span,
+                            &format!("function pointers are not allowed in const fn"));
+                        err.emit();
+                    }
+                },
+                _ => {
+                    span_bug!(self.span, "this is not allowed in const fn");
```
Comment (oli-obk) [inline src/librustc_mir/transform/qualify_consts.rs]: I think it would be strictly less ICEy if we just called `not_const` here and early returned

**Review by oli-obk — CHANGES_REQUESTED** [review]

**Comment by phansch** [comment]
I had to rebase from master which was rather complicated, so I opted to re-do this because that was quicker. I added an additional test for `const fn foo() { (||{})() }` and it's now using `self.not_const()` instead of a custom message.

**Comment by oli-obk** [comment]
@bors r+

thanks

**Review by oli-obk — APPROVED** [review]

**Pull request merged**